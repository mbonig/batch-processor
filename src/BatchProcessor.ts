import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Batcher } from './BatchProcessor/Batcher';
import { Finalizer } from './BatchProcessor/Finalizer';
import { Ingestion } from './BatchProcessor/Ingestion';
import { Processor } from './BatchProcessor/Processor';

export class BatchProcessor extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const processingTime = Duration.minutes(5);
    const ingestion = new Ingestion(this, 'Ingestion', { processingTime });
    const processor = new Processor(this, 'Processor', { queue: ingestion.queue, processingTime });
    new Batcher(this, 'Batcher', { queue: ingestion.queue, table: processor.table });
    new Finalizer(this, 'Finalizer', { table: processor.table });
  }
}
