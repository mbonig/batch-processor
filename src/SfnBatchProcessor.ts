import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

import { SfnBatcher } from './BatchProcessor/SfnBatcher';
import { SfnProcessor } from './BatchProcessor/SfnProcessor/SfnProcessor';

export class SfnBatchProcessor extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const processingTime = Duration.minutes(5);
    const batcher = new SfnBatcher(this, 'Batcher');
    new SfnProcessor(this, 'Processor', {
      processingTime,
      batcher: batcher.handler,
      bucket: batcher.bucket,
    });
  }
}
