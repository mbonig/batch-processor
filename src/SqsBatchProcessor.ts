import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Finalizer } from './BatchProcessor/Finalizer';
import { SqsBatcher } from './BatchProcessor/SqsBatcher';
import { SqsProcessor } from './BatchProcessor/SqsProcessor';

export class SqsBatchProcessor extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);
    const processingTime = Duration.minutes(5);
    const processor = new SqsProcessor(this, 'Processor', { processingTime });

    new SqsBatcher(this, 'Batcher', {
      pollingQueue: processor.pollingQueue,
      processingQueue: processor.processingQueue,
      table: processor.table,
    });
    new Finalizer(this, 'Finalizer', { queue: processor.pollingQueue, table: processor.table });

  }
}
