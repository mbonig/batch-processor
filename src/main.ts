import { App } from 'aws-cdk-lib';
import { SfnBatchProcessor } from './SfnBatchProcessor';
import { SqsBatchProcessor } from './SqsBatchProcessor';

const app = new App();

new SqsBatchProcessor(app, 'sqs-batch-processor', {
  env: {
    account: '071128183726',
    region: 'us-east-1',
  },
});

new SfnBatchProcessor(app, 'sfn-batch-processor', {
  env: {
    account: '071128183726',
    region: 'us-east-1',
  },
});

app.synth();
