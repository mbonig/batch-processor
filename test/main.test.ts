import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { BatchProcessor } from '../src/PipelineStack';

test('Snapshot', () => {
  const app = new App();
  const stack = new BatchProcessor(app, 'test');

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
