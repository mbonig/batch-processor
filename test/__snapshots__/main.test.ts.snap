// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Batcher41B281CB": {
      "DependsOn": [
        "BatcherServiceRoleDefaultPolicy1685F323",
        "BatcherServiceRole5F2E4B9A",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "d4586726caa8b9538edb5e9dd95f76e0531b35b789244be31c1b818e34aa67bf.zip",
        },
        "Environment": {
          "Variables": {
            "POLLING_QUEUE_URL": {
              "Ref": "ProcessorCheckForFinishedBCCDE528",
            },
            "PROCESSING_QUEUE_URL": {
              "Ref": "ProcessorProcessingQueueDCCDEF8D",
            },
            "TABLE_NAME": {
              "Ref": "ProcessorStateTable55BAFD12",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "BatcherServiceRole5F2E4B9A",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "BatcherServiceRole5F2E4B9A": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "BatcherServiceRoleDefaultPolicy1685F323": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorStateTable55BAFD12",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorProcessingQueueDCCDEF8D",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorCheckForFinishedBCCDE528",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "BatcherServiceRoleDefaultPolicy1685F323",
        "Roles": [
          {
            "Ref": "BatcherServiceRole5F2E4B9A",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "Finalizer4445B7FE": {
      "DependsOn": [
        "FinalizerServiceRoleDefaultPolicy8BA7CDC4",
        "FinalizerServiceRole3AF6BA9F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "f5b65e4ba03ff633ab80f6a5a4ae4b12c61bf39d1bcc47bc2e93e63195fd5ef6.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "QUEUE_URL": {
              "Ref": "ProcessorCheckForFinishedBCCDE528",
            },
            "TABLE_NAME": {
              "Ref": "ProcessorStateTable55BAFD12",
            },
          },
        },
        "Handler": "index.handler",
        "RecursiveLoop": "Allow",
        "Role": {
          "Fn::GetAtt": [
            "FinalizerServiceRole3AF6BA9F",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "FinalizerServiceRole3AF6BA9F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "FinalizerServiceRoleDefaultPolicy8BA7CDC4": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorStateTable55BAFD12",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorCheckForFinishedBCCDE528",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorCheckForFinishedBCCDE528",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "FinalizerServiceRoleDefaultPolicy8BA7CDC4",
        "Roles": [
          {
            "Ref": "FinalizerServiceRole3AF6BA9F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "FinalizerSqsEventSourcetestProcessorCheckForFinished3578DAA2855A0828": {
      "Properties": {
        "EventSourceArn": {
          "Fn::GetAtt": [
            "ProcessorCheckForFinishedBCCDE528",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "Finalizer4445B7FE",
        },
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "Processor91B1B868": {
      "DependsOn": [
        "ProcessorServiceRoleDefaultPolicy7BFE061A",
        "ProcessorServiceRoleD270C6AD",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "9f9151ddc6f56408dc9034ba83ff4efb1f2ba1c9e9d50908b116cdebc6e95126.zip",
        },
        "Environment": {
          "Variables": {
            "POLLING_QUEUE_URL": {
              "Ref": "ProcessorCheckForFinishedBCCDE528",
            },
            "PROCESSING_QUEUE_URL": {
              "Ref": "ProcessorProcessingQueueDCCDEF8D",
            },
            "TABLE_NAME": {
              "Ref": "ProcessorStateTable55BAFD12",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ProcessorServiceRoleD270C6AD",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "ProcessorCheckForFinishedBCCDE528": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": "alias/aws/sqs",
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ProcessorCheckForFinishedPolicy92A4329A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorCheckForFinishedBCCDE528",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "ProcessorCheckForFinishedBCCDE528",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "ProcessorProcessingQueueDCCDEF8D": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "KmsMasterKeyId": "alias/aws/sqs",
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "ProcessorProcessingQueuePolicyFA922EDC": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sqs:*",
              "Condition": {
                "Bool": {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorProcessingQueueDCCDEF8D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Queues": [
          {
            "Ref": "ProcessorProcessingQueueDCCDEF8D",
          },
        ],
      },
      "Type": "AWS::SQS::QueuePolicy",
    },
    "ProcessorServiceRoleD270C6AD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ProcessorServiceRoleDefaultPolicy7BFE061A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorStateTable55BAFD12",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorCheckForFinishedBCCDE528",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "sqs:SendMessage",
                "sqs:GetQueueAttributes",
                "sqs:GetQueueUrl",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorProcessingQueueDCCDEF8D",
                  "Arn",
                ],
              },
            },
            {
              "Action": "cloudwatch:PutMetricData",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "sqs:ReceiveMessage",
                "sqs:ChangeMessageVisibility",
                "sqs:GetQueueUrl",
                "sqs:DeleteMessage",
                "sqs:GetQueueAttributes",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "ProcessorProcessingQueueDCCDEF8D",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ProcessorServiceRoleDefaultPolicy7BFE061A",
        "Roles": [
          {
            "Ref": "ProcessorServiceRoleD270C6AD",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ProcessorSqsEventSourcetestProcessorProcessingQueue3F989F444998932F": {
      "Properties": {
        "BatchSize": 10,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "ProcessorProcessingQueueDCCDEF8D",
            "Arn",
          ],
        },
        "FunctionName": {
          "Ref": "Processor91B1B868",
        },
        "FunctionResponseTypes": [
          "ReportBatchItemFailures",
        ],
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "ProcessorStateTable55BAFD12": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "Region": {
              "Ref": "AWS::Region",
            },
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
        "TimeToLiveSpecification": {
          "AttributeName": "ttl",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
